### ISSUE TITLE
`cargo clippy --fix` for `collapsible_match` sometimes causes compiler error with borrow mutability

### ISSUE THREAD

**Issue by XSpielinbox** [rust-lang/rust-clippy#16903]
### Summary

Running `cargo clippy --fix` on https://github.com/cackle-rs/cackle/commit/957678acffad9e1ca4b676d12c23d0e619e79c37 results in the following error:
```
    Checking cargo-acl v0.8.0 (/home/[...]/cackle)
warning: failed to automatically apply fixes suggested by rustc to crate `cargo_acl`

after fixes were automatically applied the compiler reported errors within these files:

  * src/ui/full_term/problems_ui.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0502]: cannot borrow `self.modes` as immutable because it is also borrowed as mutable
   --> src/ui/full_term/problems_ui.rs:281:20
    |
148 |         let Some(mode) = self.modes.last_mut() else {
    |                          ---------- mutable borrow occurs here
...
281 |                 if self.modes.len() >= 2 => {
    |                    ^^^^^^^^^^          - mutable borrow later used here
    |                    |
    |                    immutable borrow occurs here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0502`.
Original diagnostics will follow.

warning: this `if` can be collapsed into the outer `match`
   --> src/config_editor.rs:98:13
    |
 98 | /             if failure.output.sandbox_config.kind != Some(SandboxKind::Disabled) {
 99 | |                 let perm_sel = PermSel::for_build_script(failure.crate_sel.pkg_name());
100 | |                 if !failure.output.sandbox_config.allow_network.unwrap_or(false) {
101 | |                     edits.push(Box::new(SandboxAllowNetwork {
...   |
105 | |                 edits.push(Box::new(DisableSandbox { perm_sel }));
106 | |             }
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
    = note: `#[warn(clippy::collapsible_match)]` on by default
help: collapse nested if block
    |
 97 ~         Problem::ExecutionFailed(failure)
 98 ~             if failure.output.sandbox_config.kind != Some(SandboxKind::Disabled) => {
 99 |                 let perm_sel = PermSel::for_build_script(failure.crate_sel.pkg_name());
...
105 |                 edits.push(Box::new(DisableSandbox { perm_sel }));
106 ~             }
    |

warning: this `if` can be collapsed into the outer `match`
   --> src/names.rs:312:29
    |
312 | / ...                   if *gt_depth == 0 {
313 | | ...                       match self.it.next() {
314 | | ...                           Some(DemangleToken::Text(text)) => {
315 | | ...                               self.it = return_point.clone();
...   |
327 | | ...                   }
    | |_______________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
311 ~                         }
312 ~                             if *gt_depth == 0 => {
313 |                                 match self.it.next() {
...
326 |                                 }
327 ~                             }
    |

warning: this `if` can be collapsed into the outer `match`
   --> src/ui/full_term/problems_ui.rs:265:17
    |
265 | /                 if !self.accept_single_enabled {
266 | |                     self.modes.push(Mode::PromptAutoAccept);
267 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
264 ~             (Mode::SelectProblem, KeyCode::Char('a'))
265 ~                 if !self.accept_single_enabled => {
266 |                     self.modes.push(Mode::PromptAutoAccept);
267 ~                 }
    |

warning: this `if` can be collapsed into the outer `match`
   --> src/ui/full_term/problems_ui.rs:282:17
    |
282 | /                 if self.modes.len() >= 2 {
283 | |                     self.modes.pop();
284 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#collapsible_match
help: collapse nested if block
    |
281 ~             (_, KeyCode::Esc)
282 ~                 if self.modes.len() >= 2 => {
283 |                     self.modes.pop();
284 ~                 }
    |

warning: `cargo-acl` (bin "cargo-acl") generated 4 warnings (run `cargo clippy --fix --bin "cargo-acl" -p cargo-acl -- --no-deps` to apply 4 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 2.71s
```

Even when updating `edition` to `2024` and `rust-version` to `1.94.1` in the project or updating to Rust nightly (nightly-x86_64-unknown-linux-gnu (913e4bea8 2026-04-22)) this error persists.

### Reproducer

Code:

see [repo](https://github.com/cackle-rs/cackle/blob/957678acffad9e1ca4b676d12c23d0e619e79c37/src/ui/full_term/problems_ui.rs#L282) as I unfortunately have not (yet) been able to find a smaller reproducer; a smaller reproducer was posted in https://github.com/rust-lang/rust-clippy/issues/16903#issuecomment-4305790116

Current output:
see above

Desired output:
no error

### Version

```text
rustc 1.95.0 (59807616e 2026-04-14)
binary: rustc
commit-hash: 59807616e1fa2540724bfbac14d7976d7e4a3860
commit-date: 2026-04-14
host: x86_64-unknown-linux-gnu
release: 1.95.0
LLVM version: 22.1.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Gri-ffin** [comment]
loosely related to #16894.

It looks like `cargo clippy --fix` rewrites this:

```rust
(_, KeyCode::Esc) => {
    if self.modes.len() >= 2 {
        self.modes.pop();
    }
}
```

into a guard:

```rust
(_, KeyCode::Esc) if self.modes.len() >= 2 => {
    self.modes.pop();
}
```
the match is built from `let Some(mode) = self.modes.last_mut() ...,` so we’re already holding an exclusive borrow of `self.modes`. After the rewrite, the guard calls `self.modes.len()` while the match is still selecting an arm, which can overlap with that borrow.

I was able to reproduce it using this:

```rust
#[derive(Clone, Copy)]
enum Mode { A, B }

struct Ui { modes: Vec<Mode> }

impl Ui {
    fn handle(&mut self, key: u8) {
        let Some(mode) = self.modes.last_mut() else { return; };

        match (mode, key) {
            (&mut Mode::A, 0) => {}

            (_, 1) => {
                if self.modes.len() >= 2 {
                    self.modes.pop();
                }
            }

            _ => {}
        }
    }
}
```

**Comment by XSpielinbox** [comment]
@Gri-ffin Thank you!

But how would I rewrite this in a way that does not cause this borrow-issue? I don't see how this is related to https://github.com/rust-lang/rust-clippy/issues/16894 except in that it applies to the same lint.

Your comment to me more reads like this may be a false positive lint of clippy?

**Comment by Gri-ffin** [comment]
it's related since in both cases `collapsible_match` suggests bad code, in #16894 it suggested malformed code, but here the code is syntactically valid but it changed the borrow/exhaustiveness behavior when turning code into a match guard. The easiest solution for now is to just `#[allow(clippy::collapsible_match)]` before the match so that `clippy --fix` ignores it. though you can try to not borrow the last element mutably if you only need to read it.

**Comment by Souradip121** [comment]
@rustbot claim

Working on this. The unsoundness comes from the if-into-guard collapse: the inner condition runs in the arm body, where NLL can already have ended the scrutinee's mutable borrow, but if it is moved into a guard it instead runs while pattern matching is still in progress and the scrutinee borrow is live. When the condition reads from the same place that the scrutinee mutably borrows (directly, or transitively through a let binding of `&mut T`), the rewritten code fails to compile.

The fix adds a precondition in front of the existing if-guard collapse path. Using `ExprUseVisitor`, it collects every place mutably borrowed by the scrutinee, with one level of expansion through `let` initializers so the let-else pattern in the reproducer (`let Some(mode) = self.modes.last_mut() else { return };`) is followed back to `self.modes`. It then collects every place read or borrowed by the inner condition and checks for overlap by base local plus projection prefix. If any pair aliases, the suggestion is suppressed. The lint still fires when the inner condition reads only places that are disjoint from the scrutinee's borrow.

Tests cover the cackle reproducer verbatim, a direct `iter_mut().next()` shape, and a positive control where the scrutinee mutably borrows a vec but the inner condition reads only an unrelated local (the lint must continue to fire there). Function-parameter `&mut T` bindings are not traced through the call boundary in this first pass; that can be tightened later if a real-world case surfaces.
